source/docs/software/wpilib-tools/robot-simulation/unit-testing.rst

```diff
@@ -0,0 +1,227 @@
+             piston.set(DoubleSolenoid.Value.kReverse);
+             motor.set(0); // turn off the motor
+           }
+
+           public void activate(double speed) {
+             if (piston.get() == DoubleSolenoid.Value.kForward) {
+               motor.set(speed);
+             } else { // if piston isn't open, do nothing
+               motor.set(0);
+             }
+           }
+
+           @Override
+           public void close() throws Exception {
+             piston.close();
+             motor.close();
+           }
+         }
+
+   .. group-tab:: C++ (Source)
+
+      .. code-block:: cpp
+
+         #include <frc2/command/SubsystemBase.h>
+         #include <frc/DoubleSolenoid.h>
+         #include <frc/PWMSparkMax.h>
+
+         #include "Constants.h"
+
+         class Intake : public frc2::SubsystemBase {
+          public:
+           void Deploy();
+           void Retract();
+           void Activate(double speed);
+
+           private:
+           frc::PWMSparkMax motor{Constants::Intake::MOTOR_PORT};
+           frc::DoubleSolenoid piston{Constants::Intake::PISTON_FWD, Constants::Intake::PISTON_REV};
+         };
+
+   .. group-tab:: C++ (Header)
+
+      .. code-block:: cpp
+
+         #include "subsystems/Intake.h"
+
+         void Intake::Deploy() {
+             piston.Set(frc::DoubleSolenoid::Value::kForward);
+         }
+
+         void Intake::Retract() {
+             piston.Set(frc::DoubleSolenoid::Value::kReverse);
+             motor.Set(0); // turn off the motor
+         }
+
+         void Intake::Activate(double speed) {
+             if (piston.Get() == frc::DoubleSolenoid::Value::kForward) {
+                 motor.Set(speed);
+             } else { // if piston isn't open, do nothing
+                 motor.Set(0);
+             }
+         }
+
+Writing Tests
+^^^^^^^^^^^^^
+
+Tests are placed inside the ``test`` source set: ``/src/test/java/`` and ``/src/test/native/`` for Java and C++ tests, respectively. In Java, each test class contains at least one test method marked with ``@org.junit.Test``, each method representing a test case. Additional methods for opening resources (such as our ``Intake`` object) before each test and closing them after are respectively marked with ``@org.junit.Before`` and ``@org.junit.After``. In C++, test fixture classes inheriting from ``testing::Test`` contain our subsystem and simulation hardware objects, and test methods are written using the ``TEST_F(testfixture, testname)`` macro. The ``SetUp()`` and ``TearDown()`` methods can be overridden in the test fixture class and will be run respectively before and after each test.
+
+Each test method should contain at least one *assertion* (``assert*()``/``EXPECT_*()``). These assertions verify a condition at runtime and fail the test if the condition isn't met. If there is more than one assertion in a test method, the first failed assertion will crash the test - execution won't reach the later assertions.
```

Comment:
``assert()`` methods won't do anything in Release mode. Did you mean the ``ASSERT_`` GTest macros? If so, you should explain what the difference between ``ASSERT_`` and ``EXPECT_`` is.

Reply:
I thought that it was clear that `assert*()` was for Java and `EXPECT_*()` was for C++.